lib/route.js

```diff
@@ -307,12 +307,30 @@ function buildRouting (options) {
         constraints.version = opts.version
       }
 
-      const headHandler = router.find('HEAD', opts.url, constraints)
-      const hasHEADHandler = headHandler != null
+      let hasHEADHandler = false
 
-      // remove the head route created by fastify
-      if (hasHEADHandler && !context[kRouteByFastify] && headHandler.store[kRouteByFastify]) {
-        router.off('HEAD', opts.url, { constraints })
+      if (shouldExposeHead && opts.method === 'GET') {
+        // We're using the "router.on" method in order to understand if there is an identical route
+        // in router, only with "HEAD" method.
+        // once "has" method will be implemented on 'find-my-way' side, we will use it.
+
+        let headHandler = null
+
+        try {
+          router.on('HEAD', opts.url, { constraints }, routeHandler, context)
+          router.off('HEAD', opts.url, constraints)
+        } catch (error) {
+          const isDuplicatedRoute = error.message.includes("Method 'HEAD' already declared for route")
+          if (isDuplicatedRoute) {
+            hasHEADHandler = true
+            headHandler = router.find('HEAD', opts.url, constraints)
+          }
+        }
+
+        // remove the head route created by fastify
+        if (hasHEADHandler && !context[kRouteByFastify] && headHandler.store[kRouteByFastify]) {
+          router.off('HEAD', headHandler.store.config.url, constraints)
+        }
```

Comment:
I'm running into constant OOM issues when trying to benchmark the solution. I imagine is because of the doubled number of routes and checks.

I'm imagining that we can do several optimizations here.

The whole check to the `HEAD` routes is duplicated, I think we can do it within this [line](https://github.com/fastify/fastify/pull/5009/files#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3R413) so we remove duplicated logic. We can even abstract it into a standalone function for better readability.
The duplicated route check should only be checked when trying to register the sibling route, not before.

Reply:
When implementing the solution I tried to keep the behavior of the code exactly the same. 
What we're doing in the `main` branch is: checking if a route exists using find > if so, under certain circumstances, remove if > set `headHandler` and `hasHEADHandler` accordingly to the rest of the code, and use them.

So, I kept it as it is, in the same location.
Also, I didn't add this check, this check exists already. The only thing that was doubled is the `router.on` `router.off` operations, which are instead of calling `router.find`.

There's a reason why this check is done right before the `router.on` that sets the route.
We can obviously move this piece of code to a different location and optimize it a bit. This could lead to bigger changes we should avoid.

But first let me ask:
1. What does the benchmark exactly do? can I see it, please?
2. Did you have this benchmark the entire time? or it's a new benchmark?
3. Did you test it on the main branch? [this line](https://github.com/fastify/fastify/pull/5009/files#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3R413) is just an if statement. Yes, there might be a way to optimize that, but I highly doubt that arithmetic operations can lead to OOM that easily.

Reply:
>What does the benchmark exactly do? can I see it, please?

It just adds a bunch of routes, so far it works on `main` but breaks on this PR:

```js
const Benchmark = require('benchmark')
const fastify = require('fastify')
const fastifypr = require('fastify-rantoledo')

const suite = new Benchmark.Suite()
const routes = Array.from({ length: 5 }).map((_, index) => `/${index + 1}`)

suite
  .add(
    'Main',
    async function () {
      const app = fastify()

      for (const route of routes) {
        app.get(route, (request, reply) => {
          reply.send('Hello World!')
        })
      }

      await app.close()
    },
    {
      minSamples: 100
    }
  )
  .add(
    'PR',
    async function () {
      const app = fastifypr()

      for (const route of routes) {
        app.get(route, (request, reply) => {
          reply.send('Hello World!')
        })
      }

      await app.close()
    },
    {
      minSamples: 100,
    }
  )
  .on('cycle', function (event) {
    console.log(String(event.target))
    global.gc()
  })
  .on('complete', function () {
    console.log('Fastest is ' + this.filter('fastest').map('name'))
  })
  .run()
```

>Note that it is important to call `await fastify.close` to deallocate the resources.

> Did you have this benchmark the entire time? or it's a new benchmark?

Its new, but we'll might want to look at it and add it soon

>Did you test it on the main branch? [this line](https://github.com/fastify/fastify/pull/5009/files#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3R413) is just an if statement. Yes, there might be a way to optimize that, but I highly doubt that arithmetic operations can lead to OOM that easily.

If there's any recursion within fmw, it is possible

Reply:
Ha... Now it worked

Reply:
Please, disregard my previous comment:
Doing it asynchronously (with `async: true` solves the issue). I imagine might be due to the GC not fully collecting the unreferenced objects on time before reaching memory limits.

So far the results were the following:
```shell
Main x 2,208 ops/sec ±0.97% (186 runs sampled)
PR x 2,228 ops/sec ±0.99% (188 runs sampled)
```

Tests:
```js
const Benchmark = require('benchmark')
const fastify = require('fastify')
const fastifypr = require('fastify-rantoledo')

const suite = new Benchmark.Suite()
const routes = Array.from({ length: 20 }).map((_, index) => `/${index + 1}`)

suite
  .add(
    'Main',
    async function () {
      const app = fastify()

      for (const route of routes) {
        app.get(route, (request, reply) => {
          reply.send('Hello World!')
        })
      }

      await app.close()
    },
    {
      minSamples: 100,
      delay: false
    }
  )
  .add(
    'PR',
    async function () {
      const app = fastifypr()

      for (const route of routes) {
        app.get(route, (request, reply) => {
          reply.send('Hello World!')
        })
      }

      await app.close()
    },
    {
      minSamples: 100,
      delay: false
    }
  )
  .on('cycle', function (event) {
    console.log(String(event.target))
    global.gc()
  })
  .on('complete', function () {
    console.log('Fastest is ' + this.filter('fastest').map('name'))
  })
  .run({
    async: true
  })
```